### PR TITLE
Update Client Service to enable Regulator for transfers

### PR DESCRIPTION
**Pull request by daveroga**
## What does this implement/fix? Explain your changes.
Update in the client to enable regulator for the transfers.

## Does this close any currently open issues?
#1372 

## What commands can I run to test the change? 
Start nightfall with `REGULATOR_URL=http://localhost:8080 NF_SERVICES_TO_START=blockchain,client,worker,optimist,deployer ./bin/start-nightfall -g -d` to indicate a regulator url for the client to create transfers enabled for the regulator.

- npx hardhat test --bail --no-compile test/unit/circuits/transfer-regulator.circom.test.mjs (This is also included in `npm run unit-test-circuits`
- npx hardhat test --bail --no-compile test/kem-dem-regulator.test.mjs (This is also included in `npm run test-general-stuff`
- npm run test-erc20-tokens (To test in real scenario you need to simulate a regulator in start-nightfall with a REGULATOR_URL  `REGULATOR_URL=http://localhost:8080 NF_SERVICES_TO_START=blockchain,client,deployer,optimist,worker ./bin/start-nightfall -g -d`)

## Any other comments?
This is part of the regulator feature and the part for the regulator service is simulated now, waiting for the corresponding PR that will be merged in this feature branch and then integrated.

### CODE REVIEW

**Review by eugypalu — APPROVED** [review]

**Review by imagobea — CHANGES_REQUESTED** [review]
Only requesting changes so that we can discuss some of the comments and take a look at the test with the SDK.

I could follow more or less the implementation till I got to the circuit part! :see_no_evil: We need to find a place to document this in a simple but comprehensive way.

Once the feature evolves, we should add more test. If possible, I would create new test suites instead of adding more cases to `erc-tokens`.

We should test that a 3rd party can open the commitment and access the data as intended. But cannot spend commitments. Etc etc.

Great work indeed!! :muscle: It is nice to see others adventuring into the circuits layer. Well done :clap:

File: docker/docker-compose.yml
Comment (daveroga) [reply docker/docker-compose.yml:61]: Done

File: nightfall-client/src/services/transfer-regulator.mjs
Comment (daveroga) [reply nightfall-client/src/services/transfer-regulator.mjs:6]: Fixed

File: nightfall-client/src/services/transfer-regulator.mjs
Comment (daveroga) [reply nightfall-client/src/services/transfer-regulator.mjs:101]: Yes, true. This was because at the beginning there was only one transfer that did the regulator feature too but finally I separated it. Fixed.

File: nightfall-client/src/services/kem-dem.mjs
Comment (daveroga) [reply nightfall-client/src/services/kem-dem.mjs:134]: Fixed. Stackoverflow :).

File: nightfall-client/src/services/kem-dem.mjs
Comment (daveroga) [reply nightfall-client/src/services/kem-dem.mjs:143]: Done

File: nightfall-client/src/services/kem-dem.mjs
Comment (daveroga) [reply nightfall-client/src/services/kem-dem.mjs:115]: It's not ephemeral in this case because it's deterministic by a combination of the senderPrivateKey and the receiverPublicKey.

**Review by imagobea — APPROVED** [review]
Go Regulator, Go!! Thanks for explaining this to me (again) :pray:

**Review by israelboudoux — CHANGES_REQUESTED** [review]

File: nightfall-client/src/services/commitment-sync.mjs
Comment (israelboudoux) [inline nightfall-client/src/services/commitment-sync.mjs:47]: I'm not sure if the transaction value is a good candidate for a being a salt/nonce. We'd need something more random. The transaction values in Nightfall are all integers, right? IMHO, they don't provide a good source of randomness.

File: nightfall-client/src/utils/computeCircuitInputs.mjs
Comment (israelboudoux) [inline nightfall-client/src/utils/computeCircuitInputs.mjs:118]: Could you please use a variable name that is more intuitive?

File: nightfall-client/src/utils/regulator.mjs
Comment (daveroga) [reply nightfall-client/src/utils/regulator.mjs:28]: Yes, the feature branch is in progress. @eugypalu  is developing this service and once this PR is merged in the feature branch we will integrate this.

File: nightfall-client/src/utils/regulator.mjs
Comment (daveroga) [reply nightfall-client/src/utils/regulator.mjs:37]: This is simulated now waiting for the service to be integrated after the merge of this PR in the feature branch.

File: nightfall-client/src/services/commitment-sync.mjs
Comment (daveroga) [reply nightfall-client/src/services/commitment-sync.mjs:47]: We store a random value in `transaction.value`. It's not the value itself of the transaction. In transfers this value is always 0 now. That's why we have taken this variable to store the nonce. Like in other parts to optimize the size of the information stored in the transaction and don't create extra fields which will increment the gas per transaction in calldata.

File: nightfall-client/src/utils/computeCircuitInputs.mjs
Comment (daveroga) [reply nightfall-client/src/utils/computeCircuitInputs.mjs:118]: This is the shared secret for the sender (analog to public ephemeral key in previous version). The sender with this shared secret only has to apply a scalarMult with his private key to get the encryption key. So I put sharedPubxxxx in the variables that only need the private of xxxx to get the encryption key. I know it's a bit difficult to understand at the beginning.

File: nightfall-client/src/services/kem-dem.mjs
Comment (israelboudoux) [inline nightfall-client/src/services/kem-dem.mjs:115]: You mention here `Observers`. Who are they? It would be good to have some details in the jsdoc or even add them to a README.

File: nightfall-client/src/utils/computeCircuitInputs.mjs
Comment (israelboudoux) [reply nightfall-client/src/utils/computeCircuitInputs.mjs:118]: I think that `sharedPubKeySender` would be better. The way it is today only by diving into the code we can grasp about.

File: test/kem-dem-regulator.test.mjs
Comment (israelboudoux) [inline test/kem-dem-regulator.test.mjs:61]: This test already exists in `test/kem-dem.test.mjs`.

File: nightfall-client/src/utils/regulator.mjs
Comment (israelboudoux) [reply nightfall-client/src/utils/regulator.mjs:28]: Got it.

File: nightfall-client/src/services/kem-dem.mjs
Comment (daveroga) [reply nightfall-client/src/services/kem-dem.mjs:115]: Done. Added to the description of the jsdoc. Observers are receiver and regulator.

File: test/kem-dem-regulator.test.mjs
Comment (daveroga) [reply test/kem-dem-regulator.test.mjs:61]: Removed duplicated test.

**Review by israelboudoux — APPROVED** [review]
Nice one @daveroga!

**Pull request merged**
